### DISCUSSION TITLE
post/18-ansible-in-aio/

### DISCUSSION THREAD

**Discussion by giscus** [lyc8503/lyc8503#34] [Announcements]
# post/18-ansible-in-aio/

前言之前我已经使用了 Ansible 管理了我一系列 VPS, 但是暂时没有使用 Ansible 管理 HomeLab. 最近我的 HomeLab 有一些硬件变更和升级, 不过由于我发现我早期的一些 Blog 比较啰嗦和无趣, 我决定把一些比较 trivial 的变更单独整合到 HomeLab Changelog 中, 所以在此不再展开介绍一年来的硬件变更. TL;DR 一下: 更换了硬盘 SN6

https://blog.lyc8503.net/post/18-ansible-in-aio/

**Comment by SuperKenVery** [comment]
不太清楚你说的nixos缺少hypevisor方案是什么意思，但根据你说的proxmox+ansible来看，你要的应该是一个方便根据nixos配置开虚拟机的虚拟机平台？

如果是这样的话，普通的nixos就可以完成这个任务，系统配置里可以配虚拟机，虚拟机就是用nix定义的。

此外，也许你可以试试在其他发行版上部署用nix定义的虚拟机，再慢慢转移到纯nixos😀

**Reply by lyc8503** [reply]
我对 Hypervisor 的首要要求是稳定, 别天天给我整幺蛾子, 那么 NixOS unstable 作为滚动发行版显然不太适合. 即使它可以很方便的回滚, 我还是需要花时间跟进上游变更才能得到安全更新, 我不希望 Hypervisor 这种基建遇到 downtime 或者带来额外的工作量影响我其他的事情进展.

NixOS stable 似乎相对用的人不多, 大家都跑去用 Flake 了, 相关资料似乎更少, 我对其支持程度表示一定的怀疑. 而且 6 个月一个版本, 新版本发了旧版本基本就 EOL, 也比较难跟上... (PVE / Debian 两个相邻版本的支持期会重叠一年) 我的预期是, 就算我很忙没空维护, Hypervisor 也应该至少稳定运行 1-2 年并得到更新.

我现在甚至花了 110 EUR/year 购买了 [Proxmox VE 的企业源订阅](https://shop.proxmox.com/index.php?rp=/store/proxmox-ve-community). (虽然初衷是支持下优秀开源软件, 但企业源是这里的副产物)

> 如果是这样的话，普通的nixos就可以完成这个任务，系统配置里可以配虚拟机，虚拟机就是用nix定义的

我搜索了一些各类方案, 包括 `microvm.nix` / `proxmox-nixos` / `NixVirt`, 都有各式各样的问题...

- microvm.nix 只能跑 NixOS 虚拟机, 我需要 Windows, 不太行
- proxmox-nixos 构建过程就遇到了不少问题, 终于修好后发现它的 WebUI 和 Nix 配置会有很多不同步的地方, 比较恶心... 排除
- [NixOS 官方 Wiki 有关虚拟化的内容](https://nixos.wiki/wiki/Virtualization) 我觉得也挺不清不楚的....

我还依赖了挺多 PVE 的功能的, 包括硬件虚拟化/直通/虚拟机快照/备份/防火墙... 将来计划搞集群, 可能还会用 Ceph 和 Replication. 我似乎没找到一个 NixOS 下功能比较完善的替代品, 都自己实现一遍工作量浩大, 效果肯定也不比 PVE. 功能缺失+软件支持缺失就是我指的"不太成熟".

> 也许你可以试试在其他发行版上部署用nix定义的虚拟机，再慢慢转移到纯nixos😀

我现在打算逐步将我的一些不那么重要 ~~(比如用来跑代理)~~ 的 VPS 全面替换成 NixOS. 看起来一份 Nix 配置可以比较方便的生成不同平台和架构的系统, 这对我来说比较方便, 有一些 LXC 容器化或者 ARM 架构的 VPS 也可以方便的跑 NixOS.

至于把 PVE 换成 NixOS 感觉道阻且长, 短期内我可能不会这么考虑...
